src/main/java/org/checkerframework/languageserver/CFTextDocumentService.java

```diff
@@ -146,21 +176,98 @@ public void didClose(DidCloseTextDocumentParams params) {
         }
     }
+
+    /**
+     * The hover request is sent from the client to the server to request hover information at a
+     * given text document position.
+     *
+     * <p>Registration Options: TextDocumentRegistrationOptions
+     */
+    @Override
+    public CompletableFuture<Hover> hover(HoverParams params) {
+        int line = params.getPosition().getLine();
+        int character = params.getPosition().getCharacter();
+        ComparablePosition currentPosition = new ComparablePosition(line, character);
+        File curFile = new File(URI.create(params.getTextDocument().getUri()));
+        RangeMap<ComparablePosition, List<String>> typeInfo = filesToTypeInfo.get(curFile);
+
+        if (typeInfo != null) {
+            List<String> rawTypeInfoForHover = typeInfo.get(currentPosition);
+            if (rawTypeInfoForHover != null) {
+                List<Either<String, MarkedString>> typeInfoForHover =
+                        rawTypeInfoForHover.stream()
+                                .map(Either::<String, MarkedString>forLeft)
+                                .collect(Collectors.toList());
+                Hover result = new Hover(typeInfoForHover);
+                return CompletableFuture.completedFuture(result);
+            }
+        }
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * Publish the given type message for the given file.
+     *
+     * @param file The file to which the type message corresponds.
+     * @param msg A type message which contains checker name, message kind, type information, and
+     *     the range of this type in the given file, separated by the delimiter ";".
+     */
+    private void publishTypeMessage(File file, String msg) {
+        if (!filesToTypeInfo.containsKey(file)) {
+            RangeMap<ComparablePosition, List<String>> rangeMap = TreeRangeMap.create();
+            filesToTypeInfo.put(file, rangeMap);
+        }
+
+        int lastDelimiter = msg.lastIndexOf(';');
+        String typeInfo = msg.substring(0, lastDelimiter);
+        String positionInfo = msg.substring(lastDelimiter + 1).trim();
+        Matcher rangeMatcher = rangePattern.matcher(positionInfo);
+        if (!rangeMatcher.matches()) {
+            throw new BugInCF("Failed to parse node position!");
+        }
+
+        int startLine = Integer.parseInt(rangeMatcher.group(1));
+        int startCol = Integer.parseInt(rangeMatcher.group(2));
+        int endLine = Integer.parseInt(rangeMatcher.group(3));
+        int endCol = Integer.parseInt(rangeMatcher.group(4));
+
+        ComparablePosition start = new ComparablePosition(startLine, startCol);
+        ComparablePosition end = new ComparablePosition(endLine, endCol);
+
+        RangeMap<ComparablePosition, List<String>> currentTypeInfo = filesToTypeInfo.get(file);
+        List<String> typeInfoForPosition = currentTypeInfo.get(start);
```

Comment:
```suggestion
        int lastDelimiter = msg.lastIndexOf(';');
        String typeInfo = msg.substring(0, lastDelimiter);
        String positionInfo = msg.substring(lastDelimiter + 1).trim();
        Matcher rangeMatcher = rangePattern.matcher(positionInfo);
        if (!rangeMatcher.matches()) {
            throw new BugInCF("Failed to parse node position!");
        }

        int startLine = Integer.parseInt(rangeMatcher.group(1));
        int startCol = Integer.parseInt(rangeMatcher.group(2));
        int endLine = Integer.parseInt(rangeMatcher.group(3));
        int endCol = Integer.parseInt(rangeMatcher.group(4));

        ComparablePosition start = new ComparablePosition(startLine, startCol);
        ComparablePosition end = new ComparablePosition(endLine, endCol);

        RangeMap<ComparablePosition, List<String>> currentTypeInfo = filesToTypeInfo.get(file);
        if (currentTypeInfo == null) {
            currentTypeInfo = TreeRangeMap.create();
            filesToTypeInfo.put(file, currentTypeInfo);
        }

        List<String> typeInfoForPosition = currentTypeInfo.get(start);
```

Reply:
The only change here is to move the `filesToTypeInfo.get(file)` check from the beginning to where it is used, putting related code next to each other, making clear why `currentTypeInfo` is guaranteed to be non-null.
Maybe do the edit yourself, as I just did it in the webeditor here.